### DISCUSSION TITLE
OTA micropython firmware update possible ?

### DISCUSSION THREAD

**Discussion by ondrej1024** [micropython/micropython#14187] [RP2040 / Pico]
I am trying to find out if it is possible to update the MP firmware on the Pico-W (or Pico with Ethernet) via network connection. 
I found some tools which can update files in the filesystem via network. But I am looking for a way to replace the whole MP firmware including the frozen modules baked into the firmware. It's what you would do with `mpremote` when you load a `uf2` file vie USB connection to the Pico, but via network connection.

Is this something that can be done or has even already been implemented?

**Comment by jimmo** [comment]
@ondrej1024 You'd need a second-stage bootloader to achieve this (as the built-in bootloader doesn't support this).

For example on the stm32 port, MicroPython provides an optional second-stage bootloader (`mboot`) that can read the MicroPython filesystem to apply a compressed firmware image. (Including cryptographic signature verification).

Unfortunately I don't know of any existing projects that provide this functionality for rp2040 (although I haven't looked closely).

**Reply by ondrej1024** [reply]
Thanks @jimmo 
I had a look around and there seem to be at least a couple of projects which implement a second-stage bootloader for the RP2040:

- [picowota](https://github.com/usedbytes/picowota)
- [pico-flashloader](https://github.com/rhulme/pico-flashloader)

So I guess this would have to be integrated into MicroPython. But I'm afraid we won't see this anytime soon :roll_eyes:
